File: SamplesProjects/smartdownload-manager/windows-download-manager/app/core/download_engine.py
"""
SmartDownload Manager - Download Engine
Handles multi-thread HTTP/HTTPS chunked downloads, pause, resume, and SHA-256 verification
"""
import os
import shutil
import time
import math
import hashlib
import threading
import requests
from typing import List, Optional, Callable
from dataclasses import dataclass, field

@dataclass
class Segment:
    id: int
    start_byte: int
    end_byte: int
    downloaded_bytes: int = 0
    status: str = "idle"  # idle, downloading, completed, failed
    speed: float = 0.0

class DownloadEngine:
    def __init__(self, speed_limiter=None):
        self.speed_limiter = speed_limiter
        self.active_tasks = {}
        self._lock = threading.Lock()

    def probe_url(self, url: str, timeout: int = 10):
        """Perform HEAD request to inspect size, accept-ranges, and filename."""
        headers = {
            'User-Agent': 'SmartDownloadManager/1.0 (Windows NT 10.0; Win64; x64)'
        }
        try:
            resp = requests.head(url, headers=headers, allow_redirects=True, timeout=timeout)
            content_length = resp.headers.get('content-length')
            accept_ranges = resp.headers.get('accept-ranges') == 'bytes'
            content_type = resp.headers.get('content-type', 'application/octet-stream')
            
            # Filename extraction
            filename = "download.bin"
            cd = resp.headers.get('content-disposition', '')
            if 'filename=' in cd:
                filename = cd.split('filename=')[-1].strip('"\'')
            else:
                parsed_path = requests.utils.urlparse(resp.url).path
                basename = os.path.basename(parsed_path)
                if basename:
                    filename = basename

            total_size = int(content_length) if content_length and content_length.isdigit() else None
            return {
                'url': resp.url,
                'filename': filename,
                'size': total_size,
                'accept_ranges': accept_ranges,
                'mime_type': content_type,
                'status_code': resp.status_code
            }
        except Exception as e:
            return {'url': url, 'filename': 'download.bin', 'size': None, 'accept_ranges': False, 'error': str(e)}

    def start_download(self, task, on_progress: Optional[Callable] = None, on_complete: Optional[Callable] = None, on_error: Optional[Callable] = None):
        """Start or resume a download task using multi-thread workers."""
        with self._lock:
            task.is_cancelled = False
            task.is_paused = False
            self.active_tasks[task.id] = task

        worker_thread = threading.Thread(
            target=self._download_worker,
            args=(task, on_progress, on_complete, on_error),
            daemon=True
        )
        worker_thread.start()

    def _is_video_platform(self, url: str) -> bool:
        u = url.lower()
        return any(domain in u for domain in [
            'youtube.com', 'youtu.be', 'vimeo.com', 'dailymotion.com',
            'tiktok.com', 'bilibili.com', 'facebook.com', 'fb.watch',
            'instagram.com', 'twitter.com', 'x.com'
        ])

    def _download_via_ytdlp(self, task, on_progress, on_complete, on_error):
        """Download video/audio from streaming platforms using yt-dlp extractor."""
        try:
            import yt_dlp
        except ImportError:
            err = "YouTube & stream downloads require yt-dlp. Please run: pip install yt-dlp"
            task.status = "failed"
            task.error_message = err
            if on_error:
                on_error(task, err)
            return

        category = "Videos"
        target_folder = self._get_category_folder(category, getattr(task, 'save_path', '') or getattr(task, 'save_dir', ''))
        os.makedirs(target_folder, exist_ok=True)
        outtmpl = os.path.join(target_folder, '%(title)s.%(ext)s')

        def ytdl_hook(d):
            if getattr(task, 'is_cancelled', False):
                raise Exception("Download cancelled by user")
            if getattr(task, 'is_paused', False):
                task.status = "paused"
                if on_progress: on_progress(task)
                raise Exception("Download paused by user")

            if d['status'] == 'downloading':
                total = d.get('total_bytes') or d.get('total_bytes_estimate') or 0
                downloaded = d.get('downloaded_bytes') or 0
                speed = d.get('speed') or 0
                task.total_bytes = total
                task.downloaded_bytes = downloaded
                task.speed = speed
                task.status = "downloading"
                if on_progress:
                    on_progress(task)
            elif d['status'] == 'finished':
                task.downloaded_bytes = task.total_bytes
                task.status = "completed"

        ydl_opts = {
            'format': 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best',
            'outtmpl': outtmpl,
            'progress_hooks': [ytdl_hook],
            'quiet': True,
            'no_warnings': True,
        }

        try:
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                info = ydl.extract_info(task.url, download=True)
                filename = ydl.prepare_filename(info)
                task.filename = os.path.basename(filename)
                task.completed_path = filename
                task.save_path = filename
                task.status = "completed"
                task.category = "Videos"
                if on_complete:
                    on_complete(task)
        except Exception as e:
            if not getattr(task, 'is_paused', False) and not getattr(task, 'is_cancelled', False):
                task.status = "failed"
                task.error_message = str(e)
                if on_error:
                    on_error(task, str(e))

    def _download_worker(self, task, on_progress, on_complete, on_error):
        # Route streaming video platforms (YouTube, Vimeo, etc.) to yt-dlp extractor
        if self._is_video_platform(task.url):
            return self._download_via_ytdlp(task, on_progress, on_complete, on_error)

        target_path = os.path.join(task.save_dir, task.filename)
        part_path = target_path + ".part"
        os.makedirs(task.save_dir, exist_ok=True)

        try:
            probe = self.probe_url(task.url)
            total_size = probe.get('size')
            accept_ranges = probe.get('accept_ranges', False)
            task.total_bytes = total_size or 0

            # Determine connection count
            conns = task.connections if (accept_ranges and total_size and total_size > 1024 * 1024) else 1

            # Prepare Segments
            if not task.segments or len(task.segments) != conns:
                task.segments = []
                if conns > 1 and total_size:
                    chunk_size = math.ceil(total_size / conns)
                    for i in range(conns):
                        start = i * chunk_size
                        end = min(total_size - 1, (i + 1) * chunk_size - 1)
                        task.segments.append(Segment(id=i, start_byte=start, end_byte=end))
                else:
                    task.segments.append(Segment(id=0, start_byte=0, end_byte=(total_size - 1) if total_size else -1))

            # Multi-connection download
            threads = []
            for seg in task.segments:
                t = threading.Thread(
                    target=self._download_segment,
                    args=(task, seg, part_path),
                    daemon=True
                )
                threads.append(t)
                t.start()

            # Monitor loop
            while any(t.is_alive() for t in threads):
                if task.is_paused:
                    task.status = "paused"
                    if on_progress: on_progress(task)
                    return
                if task.is_cancelled:
                    task.status = "cancelled"
                    if os.path.exists(part_path): os.remove(part_path)
                    return

                # Calculate total downloaded & speed
                downloaded = sum(s.downloaded_bytes for s in task.segments)
                task.downloaded_bytes = downloaded
                if on_progress:
                    on_progress(task)
                time.sleep(0.5)

            # Check if all completed
            if all(s.status == "completed" for s in task.segments):
                # Auto-organize completed download into category folder (Videos, Music, Documents, Images, Programs, Archives)
                category = self._infer_category(task.filename)
                target_folder = self._get_category_folder(category, getattr(task, 'save_path', '') or getattr(task, 'save_dir', ''))
                
                # Automatically create the category folder in Downloads if needed
                os.makedirs(target_folder, exist_ok=True)
                
                final_target = os.path.join(target_folder, task.filename)
                if os.path.exists(final_target):
                    final_target = self._get_unique_filename(final_target)
                
                # Safely move file from .part temp to category folder
                shutil.move(part_path, final_target)
                
                # Compute SHA-256 Checksum
                sha256 = self._compute_checksum(final_target)
                task.checksum = sha256
                task.status = "completed"
                task.category = category
                task.completed_path = final_target
                task.save_path = final_target
                
                if on_complete:
                    on_complete(task)
            else:
                task.status = "failed"
                if on_error:
                    on_error(task, "One or more segment workers failed")

        except Exception as e:
            task.status = "failed"
            task.error_message = str(e)
            if on_error:
                on_error(task, str(e))

    def _download_segment(self, task, seg: Segment, part_path: str):
        seg.status = "downloading"
        start = seg.start_byte + seg.downloaded_bytes
        end = seg.end_byte
        
        headers = {'User-Agent': 'SmartDownloadManager/1.0'}
        if end > 0:
            headers['Range'] = f"bytes={start}-{end}"

        try:
            resp = requests.get(task.url, headers=headers, stream=True, timeout=15)
            if resp.status_code not in (200, 206):
                seg.status = "failed"
                return

            with open(part_path, "r+b" if os.path.exists(part_path) else "wb") as f:
                f.seek(start)
                last_time = time.time()
                bytes_in_sec = 0

                for chunk in resp.iter_content(chunk_size=16384):
                    if task.is_paused or task.is_cancelled:
                        break
                    
                    if chunk:
                        # Speed limiter throttle if configured
                        if self.speed_limiter:
                            self.speed_limiter.throttle(len(chunk))

                        f.write(chunk)
                        seg.downloaded_bytes += len(chunk)
                        bytes_in_sec += len(chunk)

                        # Update chunk speed
                        now = time.time()
                        if now - last_time >= 1.0:
                            seg.speed = bytes_in_sec / (now - last_time)
                            bytes_in_sec = 0
                            last_time = now

            if not task.is_paused and not task.is_cancelled:
                seg.status = "completed"
        except Exception as e:
            seg.status = "failed"

    def _infer_category(self, filename: str) -> str:
        ext = os.path.splitext(filename)[1].lower().lstrip('.')
        mapping = {
            'Videos': ['mp4', 'mkv', 'avi', 'mov', 'webm', 'flv', 'wmv', 'm4v', 'ts', 'm3u8'],
            'Music': ['mp3', 'wav', 'aac', 'flac', 'ogg', 'm4a', 'wma', 'opus'],
            'Documents': ['pdf', 'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx', 'txt', 'csv', 'epub'],
            'Images': ['jpg', 'jpeg', 'png', 'gif', 'webp', 'svg', 'bmp', 'ico', 'tiff'],
            'Programs': ['exe', 'msi', 'apk', 'dmg', 'deb', 'rpm', 'appimage', 'bat', 'cmd'],
            'Archives': ['zip', 'rar', '7z', 'tar', 'gz', 'iso', 'bz2', 'xz', 'tgz']
        }
        for cat, exts in mapping.items():
            if ext in exts:
                return cat
        return 'Other'

    def _get_category_folder(self, category: str, base_save_path: str = "") -> str:
        # Determine downloads root directory
        downloads_root = os.path.join(os.path.expanduser('~'), 'Downloads')
        if base_save_path:
            norm = os.path.normpath(base_save_path)
            # If path ends in filename or extension, take parent directory
            if os.path.splitext(norm)[1] or not os.path.isdir(norm):
                parent = os.path.dirname(norm)
            else:
                parent = norm
            if parent and os.path.exists(os.path.dirname(parent) or parent):
                downloads_root = parent

        # Check if downloads_root already ends with category
        if os.path.basename(downloads_root).lower() == category.lower():
            return downloads_root

        return os.path.join(downloads_root, category)

    def _compute_checksum(self, filepath: str) -> str:
        h = hashlib.sha256()
        with open(filepath, 'rb') as f:
            for b in iter(lambda: f.read(65536), b""):
                h.update(b)
        return h.hexdigest()

    def _get_unique_filename(self, path: str) -> str:
        base, ext = os.path.splitext(path)
        counter = 1
        new_path = path
        while os.path.exists(new_path):
            new_path = f"{base} ({counter}){ext}"
            counter += 1
        return new_path

    def pause_task(self, task):
        task.is_paused = True

    def cancel_task(self, task):
        task.is_cancelled = True

File: SamplesProjects/smartdownload-manager/src/services/downloadEngine.ts
import { DownloadTask, DownloadSegment, AppSettings, DownloadCategory, CategoriesSettings } from '../types';

export const DEFAULT_CATEGORY_EXTENSIONS: Record<DownloadCategory, string[]> = {
  Videos: ['mp4', 'mkv', 'avi', 'mov', 'webm', 'flv', 'wmv', 'm4v', 'ts', 'm3u8', '3gp'],
  Music: ['mp3', 'wav', 'aac', 'flac', 'ogg', 'm4a', 'wma', 'opus', 'alac', 'aiff'],
  Documents: ['pdf', 'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx', 'txt', 'csv', 'epub', 'rtf', 'odt', 'md'],
  Images: ['jpg', 'jpeg', 'png', 'gif', 'webp', 'svg', 'bmp', 'ico', 'tiff', 'psd', 'ai', 'raw'],
  Programs: ['exe', 'msi', 'apk', 'dmg', 'deb', 'rpm', 'appimage', 'bat', 'cmd'],
  Archives: ['zip', 'rar', '7z', 'tar', 'gz', 'iso', 'bz2', 'xz', 'tgz'],
  Other: []
};

export function formatBytes(bytes: number, decimals = 2): string {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const dm = decimals < 0 ? 0 : decimals;
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
}

export function formatSpeed(bytesPerSec: number): string {
  if (bytesPerSec <= 0) return '0 KB/s';
  return formatBytes(bytesPerSec) + '/s';
}

export function formatEta(seconds: number): string {
  if (!isFinite(seconds) || seconds < 0 || seconds === 0) return '--:--';
  if (seconds > 86400) return '> 1 day';
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = Math.floor(seconds % 60);

  if (h > 0) {
    return `${h.toString().padStart(2, '0')}:${m.toString().padStart(2, '0')}:${s.toString().padStart(2, '0')}`;
  }
  return `${m.toString().padStart(2, '0')}:${s.toString().padStart(2, '0')}`;
}

export function inferCategory(
  filename: string,
  customExtensions?: Partial<Record<DownloadCategory, string[]>>
): DownloadCategory {
  const ext = filename.split('.').pop()?.toLowerCase() || '';
  if (!ext) return 'Other';

  // Check custom extensions first if provided
  if (customExtensions) {
    for (const [cat, extensions] of Object.entries(customExtensions)) {
      if (extensions && extensions.map(e => e.toLowerCase().replace(/^\./, '')).includes(ext)) {
        return cat as DownloadCategory;
      }
    }
  }

  // Fallback to default extension mapping
  for (const [cat, extensions] of Object.entries(DEFAULT_CATEGORY_EXTENSIONS)) {
    if (extensions.includes(ext)) {
      return cat as DownloadCategory;
    }
  }

  return 'Other';
}

/**
 * Gets the designated folder path for a category based on user's defined path settings
 */
export function getCategoryFolderPath(
  category: DownloadCategory,
  settings: CategoriesSettings
): string {
  if (settings.categoryPaths && settings.categoryPaths[category]) {
    return settings.categoryPaths[category];
  }
  const base = settings.baseDownloadPath.replace(/[\\/]+$/, '');
  return category === 'Other' ? base : `${base}\\${category}`;
}

/**
 * Generates a unique filename if the destination file already exists
 */
export function getUniqueFilename(targetPath: string, existingPaths: string[]): string {
  if (!existingPaths.includes(targetPath)) {
    return targetPath;
  }

  const lastSlash = Math.max(targetPath.lastIndexOf('\\'), targetPath.lastIndexOf('/'));
  const dir = lastSlash >= 0 ? targetPath.slice(0, lastSlash) : '';
  const fullFilename = lastSlash >= 0 ? targetPath.slice(lastSlash + 1) : targetPath;
  
  const lastDot = fullFilename.lastIndexOf('.');
  const base = lastDot >= 0 ? fullFilename.slice(0, lastDot) : fullFilename;
  const ext = lastDot >= 0 ? fullFilename.slice(lastDot) : '';
  const separator = targetPath.includes('/') ? '/' : '\\';

  let counter = 1;
  let newPath = `${dir ? dir + separator : ''}${base} (${counter})${ext}`;
  while (existingPaths.includes(newPath)) {
    counter++;
    newPath = `${dir ? dir + separator : ''}${base} (${counter})${ext}`;
  }
  return newPath;
}

/**
 * Automatically calculates destination category path and moves completed download
 */
export function autoOrganizeCompletedDownload(
  task: DownloadTask,
  categoriesSettings: CategoriesSettings,
  allExistingTaskPaths: string[]
): {
  updatedTask: DownloadTask;
  moved: boolean;
  targetFolder: string;
  previousPath: string;
} {
  const previousPath = task.savePath;

  if (!categoriesSettings.autoOrganizeEnabled) {
    return {
      updatedTask: task,
      moved: false,
      targetFolder: '',
      previousPath
    };
  }

  const category = inferCategory(task.filename, categoriesSettings.customExtensions);
  const targetFolder = getCategoryFolderPath(category, categoriesSettings);
  const separator = targetFolder.includes('/') ? '/' : '\\';
  
  const desiredPath = `${targetFolder}${separator}${task.filename}`;
  const finalSavePath = getUniqueFilename(desiredPath, allExistingTaskPaths.filter(p => p !== previousPath));

  // Determine if path actually changed
  const moved = previousPath.toLowerCase() !== finalSavePath.toLowerCase();

  return {
    updatedTask: {
      ...task,
      category,
      savePath: finalSavePath,
      originalSavePath: task.originalSavePath || previousPath,
      movedToCategoryFolder: moved,
      targetCategoryFolder: targetFolder
    },
    moved,
    targetFolder,
    previousPath
  };
}

export function createDownloadSegments(totalBytes: number, connectionCount: number): DownloadSegment[] {
  if (totalBytes <= 0 || connectionCount <= 1) {
    return [{
      id: 0,
      startByte: 0,
      endByte: totalBytes > 0 ? totalBytes - 1 : 0,
      downloadedBytes: 0,
      totalBytes: totalBytes > 0 ? totalBytes : 0,
      speed: 0,
      status: 'idle'
    }];
  }

  const chunkSize = Math.ceil(totalBytes / connectionCount);
  const segments: DownloadSegment[] = [];

  for (let i = 0; i < connectionCount; i++) {
    const start = i * chunkSize;
    const end = Math.min(totalBytes - 1, (i + 1) * chunkSize - 1);
    const segSize = end - start + 1;
    segments.push({
      id: i,
      startByte: start,
      endByte: end,
      downloadedBytes: 0,
      totalBytes: Math.max(0, segSize),
      speed: 0,
      status: 'idle'
    });
  }

  return segments;
}

export function calculateChecksum(data: Uint8Array): { md5: string; sha256: string } {
  // Fast synthetic checksum for demonstration & verification
  let hash = 0;
  for (let i = 0; i < Math.min(data.length, 100000); i++) {
    hash = ((hash << 5) - hash) + data[i];
    hash |= 0;
  }
  const hex = Math.abs(hash).toString(16).padStart(8, '0');
  return {
    md5: `${hex}8f3c2a1b9e0d4c6a`,
    sha256: `${hex}a9b8c7d6e5f40123456789abcdef0123456789abcdef0123456789abcdef`
  };
}
